### ISSUE TITLE
Refactoring: Use `Processor` + `DefaultSequenceIterator` in `Generator`

### ISSUE THREAD

**Issue by georgwiese** [powdr-labs/powdr#590]
We currently have significant code duplication between `BlockMachine` (which uses the `Processor` type once #528 is merged) and `Generator`, which implements its own way of processing rows. So I looked into what it would take to use the `Processor` + `DefaultSequenceIterator` approach in `Generator` as well.

#### Performance optimizations:
  - Split identities by next ref: For each row, first go over identities without a next ref, then identities with a next ref
    - -> WIP (#603)
  - Concept of incomplete identities: If an identity is complete, don't process it again in the current row
    - -> WIP (#603)
  - Loop detection + row proposal mechanism

#### Different assumptions for machine types:
  - Directions:
    - Block machines are processed forward, backward, and forward again
    - VMs are just processed forward once
    - -> WIP (#603)
  - Handle non-static block size
    - -> WIP (#603)

#### Better error reporting:
  - After visiting a row for the last time, assert that all incomplete identities pass, assuming 0 for unknown values

#### Additional features:
- Process witness queries

## Proposal

This is a way to generalize the `Processor` + `DefaultSequenceIterator` approach to support all features needed for the `Generator`.

#### 1. Extend the actions yielded by `DefaultSequenceIterator`

This could look like this:
```rust
// Previously called `IdentityInSequence`
pub enum Action {
    /// Process identity of a given index
    Process(usize),
    /// Check identity on a given row, assuming zero for unknown values
    Check(usize)
    /// Go over all columns with a query and process it
    Queries,
    /// Process the outer query
    OuterQuery,
}

pub struct SequenceStep {
    /// Which row to operate on
    pub row_index: usize,
    /// Which action to take
    pub identity: Action,
}
```

Notes:
- `Process` (with different name) and `OuterQuery` exist currently.
- `Check` would be yielded for every incomplete identity whenever we visit a given row for the last time. This would add a slight overhead to the block machine but also fail in cases that wouldn't be detected now.
- `Queries` is yielded once after each pass through the identities. This would give block machines the ability to have witness queries!

#### 2. Instantiate the sequence iterator

The constructor arguments would be:
- `directions`:
  - block machine: forward, backward, forward
  - VM: forward
- `identities_count`: Same as now

`directions` is new, `block_size` and `outer_query_row` are dropped (see the next step).

#### 3. Pass more information to the sequence iterator

Currently we have a `DefaultSequenceIterator::report_progress(bool)` function, so that the iterator knows when to move on to the next row. We should have a similar function that passes more information:
- `progress: bool` -> Same as before; if `true`, will do another pass over all (incomplete) identities
- `identity_complete: bool` -> If `true`, won't yield same identity in current row
- `on_outer_query_row: bool` -> Add outer query to current row
- `on_last_row: bool` -> Finish or reverse direction

Notes:
- `on_outer_query_row` and `on_last_row` are currently the same in practice (always `false` for `Generator` and `true` in `BlockMachine` on the last row), so it could be one flag. In future it might make sense to distinguish the two cases though.
- An unfortunate consequence of this design is that `on_outer_query_row` needs to be set to `true` before the iterator can yield the outer query, so it wouldn't be possible to yield it as the first identity (which we don't do now)
- The upside of this approach (over specifying the outer query row in advance like now) is that in the future we can support secondary VMs (the processor would pass `true` if the latch is `1`), for which we don't know the row in advance

#### 4. Changes to `Processor`

Besides implementing the new actions and the new `report_progress` call, it would need to:
- Instead of working on a static list of rows, it would have to dynamically append new rows if needed
- Sort identities once such that identities with next reference go last. This way, things work like now, and the sequence iterator doesn't have to be aware of that.
- Implement loop detection + row proposal mechanism.

**Comment by leonardoalt** [comment]
@georgwiese was this done in the meantime?

**Comment by georgwiese** [comment]
I think I gave up on the approach described in the issue, but the general idea is now implemented: There is a `VmProcessor` which uses a `Processor` (as does `BlockProcessor`). `BlockProcessor` is a thin wrapper around a `ProcessingSequenceIterator` + `Processor`, but `VmProcessor` does not use `ProcessingSequenceIterator`.

Describing the VM solving algorithm as an iterator was too painful. We should revisit once Rust supports Generators ;)
